internal/app/go-filecoin/storage_market_connector/client.go

```diff
@@ -0,0 +1,229 @@
+		if err != nil {
+			return err
+		}
+
+		minerAddr, err := address.NewFromString(minerAddrStr)
+		if err != nil {
+			return err
+		}
+
+		info := mState.Info
+		infos = append(infos, &storagemarket.StorageProviderInfo{
+			Address:    minerAddr,
+			Owner:      info.Owner,
+			Worker:     info.Worker,
+			SectorSize: uint64(info.SectorSize),
+			PeerID:     info.PeerId,
+		})
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return infos, nil
+}
+
+// ValidatePublishedDeal validates a deal has been published correctly
+// Adapted from https://github.com/filecoin-project/lotus/blob/3b34eba6124d16162b712e971f0db2ee108e0f67/markets/storageadapter/client.go#L156
+func (s *StorageClientNodeConnector) ValidatePublishedDeal(ctx context.Context, deal storagemarket.ClientDeal) (uint64, error) {
+	// Fetch receipt to return dealId
+	waitCtx, cancel := context.WithTimeout(ctx, 10*time.Millisecond)
+	defer cancel()
+
+	var publishMsg *types.SignedMessage
+	var publishReceipt *types.MessageReceipt
+
+	err := s.waiter.Wait(waitCtx, *deal.PublishMessage, func(block *block.Block, signedMessage *types.SignedMessage, receipt *types.MessageReceipt) error {
+		publishMsg = signedMessage
+		publishReceipt = receipt
+		return nil
+	})
+
+	if err != nil {
+		return 0, err
+	}
+
+	unsigned := publishMsg.Message
+
+	minerWorker, err := s.getFCWorker(ctx, deal.Proposal.Provider)
+	if err != nil {
+		return 0, err
+	}
+
+	if unsigned.From != minerWorker {
+		return 0, xerrors.Errorf("deal wasn't published by storage provider: from=%s, provider=%s", unsigned.From, deal.Proposal.Provider)
+	}
+
+	if unsigned.To != fcaddr.StorageMarketAddress {
+		return 0, xerrors.Errorf("deal publish message wasn't set to StorageMarket actor (to=%s)", unsigned.To)
+	}
+
+	if unsigned.Method != fcsm.PublishStorageDeals {
+		return 0, xerrors.Errorf("deal publish message called incorrect method (method=%s)", unsigned.Method)
+	}
+
+	values, err := abi.DecodeValues(unsigned.Params, []abi.Type{abi.StorageDealProposals})
+	if err != nil {
+		return 0, err
+	}
+
+	msgProposals := values[0].Val.([]types.StorageDealProposal)
+
+	proposal := msgProposals[0] // TODO: Support more than one deal
+
+	// TODO: find a better way to do this
+	equals := bytes.Equal(proposal.PieceRef, deal.Proposal.PieceRef) &&
+		uint64(proposal.PieceSize) == deal.Proposal.PieceSize &&
+		//proposal.Client == deal.Proposal.Client &&
+		//proposal.Provider == deal.Proposal.Provider &&
+		uint64(proposal.ProposalExpiration) == deal.Proposal.ProposalExpiration &&
+		uint64(proposal.Duration) == deal.Proposal.Duration &&
+		uint64(proposal.StoragePricePerEpoch) == deal.Proposal.StoragePricePerEpoch.Uint64() &&
+		uint64(proposal.StorageCollateral) == deal.Proposal.StorageCollateral.Uint64() &&
+		bytes.Equal([]byte(*proposal.ProposerSignature), deal.Proposal.ProposerSignature.Data)
+
+	if equals {
+		return leb128.ToUInt64(publishReceipt.Return[0]), nil // TODO: Is this correct?
```

Comment:
You should probably be using the abi package functions to decode this (though maybe I'm wrong that receipts are abi encoded?) or otherwise the encoding package.  Casting to leb128 varint is almost certainly wrong for where we are headed because big ints won't be encoded in leb128 anymore (and isn't this supposed to be a little int?)

Reply:
done